xla/backends/gpu/runtime/nvshmem_get_thunk.cc

```diff
@@ -0,0 +1,170 @@
+      hlo_name_(instr->name()) {}
+
+absl::Status NvshmemGetThunk::Initialize(const InitializeParams& params) {
+  TF_RETURN_IF_ERROR(NvshmemCollectiveThunk::Initialize(params));
+  if (execution_counters_) {
+    TF_RETURN_IF_ERROR(execution_counters_->Initialize(
+        params.executor, params.collective_params->run_id));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status NvshmemGetThunk::RunNvshmemCollective(const ExecuteParams& params,
+                                                   se::Stream& stream) {
+  TF_ASSIGN_OR_RETURN(auto* collectives, GetNvshmemCollectivesFromRegistry());
+  TF_ASSIGN_OR_RETURN(std::unique_ptr<Communicator> nvshmem_comm,
+                      collectives->CreateCommunicator());
+  TF_ASSIGN_OR_RETURN(
+      std::vector<DeviceBufferPair> device_buffers,
+      ConvertToDeviceBuffers(params, {buffer_},
+                             config_.config.operand_element_type));
+  TF_RET_CHECK(device_buffers.size() == 1) << "Expected one buffer pair.";
+
+  GlobalDeviceId global_device_id = params.collective_params->global_device_id;
+
+  TF_ASSIGN_OR_RETURN(const DeviceAssignment::LogicalID current_logical_id,
+                      params.collective_params->device_assn->LogicalIdForDevice(
+                          global_device_id));
+  const int64_t current_id =
+      config_.config.group_mode == CollectiveOpGroupMode::kCrossReplica
+          ? current_logical_id.replica_id
+          : current_logical_id.computation_id;
+  std::string device_string =
+      CollectiveThunk::GetDeviceString(*params.collective_params);
+
+  int device_ordinal = stream.parent()->device_ordinal();
+  const NvshmemP2PConfig::SourceTargetMapEntry source_target =
+      NvshmemP2PConfig::GetSourceTarget(config_.id_to_source_target,
+                                        device_ordinal);
+  DeviceBufferPair& buffer = device_buffers[0];
+
+  VLOG(3) << "Performing Get from device ordinal: " << device_ordinal
+          << ", global_id: " << global_device_id
+          << ", current_id: " << current_id << ", group mode: "
+          << CollectiveOpGroupModeToString(config_.config.group_mode) << " ("
+          << hlo_name_ << ")";
+
+  const std::optional<int64_t> source_id = source_target.source;
+
+  VLOG(3) << absl::StreamFormat("%s : id = %d, source_id = %d", device_string,
+                                current_id, source_id.value_or(-1));
+
+  // For receive operations, we need to register the buffer with NVSHMEM
+  // and store it in the global struct for the send operation to use
+  if (source_id.value_or(-1) == -1) {
+    VLOG(3) << "Storing destination device " << device_ordinal << " and buffer "
+            << buffer.destination_buffer.opaque();
+    g_nvshmem_buffer_addresses.StoreNvshmemPtr(
+        device_ordinal, buffer.destination_buffer.opaque());
+  } else {
+    VLOG(3) << "No source ID found, not storing buffer information";
+  }
+
+  // Get source buffer from source peer if needed.
+  if (source_id) {
```

Comment:
Prefer early returns to avoid deeply nested if else branches. I.e.
```c++
if (!source_id) {
  VLOG() << ...;
  return absl::OkStatus();
}

if (!should_run) {
  VLOG() << ...;
  return absl::OkStatus();
}

...
```

Reply:
There is a difference between `!source` and `source_id.value_or(-1) == -1`. Simplified the code as much as possible.